cadc-dali-parquet/src/main/java/ca/nrc/cadc/dali/tables/parquet/ParquetReader.java

```diff
@@ -69,199 +69,248 @@
 import org.apache.parquet.hadoop.ParquetFileReader;
 import org.apache.parquet.hadoop.metadata.ParquetMetadata;
 import org.apache.parquet.io.InputFile;
-import org.apache.parquet.io.SeekableInputStream;
+import org.apache.parquet.schema.GroupType;
+import org.apache.parquet.schema.LogicalTypeAnnotation;
 import org.apache.parquet.schema.MessageType;
+import org.apache.parquet.schema.PrimitiveType;
 import org.apache.parquet.schema.Type;
 
-
-//TODO: Temporary implementation
 public class ParquetReader {
 
     private static final Logger log = Logger.getLogger(ParquetReader.class);
 
-    public TableShape read(InputStream inputStream) throws IOException {
+    private static final long BYTE_LIMIT = 1024 * 1024L; // 1 MiB
 
-        InputFile inputFile = getInputFile(inputStream);
-        int recordCount = 0;
-        int columnCount = 0;
+    private static final FormatFactory formatFactory = new FormatFactory();
 
-        ParquetMetadata metadata = ParquetFileReader.open(inputFile).getFooter();
-        MessageType schema = metadata.getFileMetaData().getSchema();
-        columnCount = schema.getFieldCount();
+    public VOTableDocument read(RandomAccessFile file) throws IOException {
+        ParquetInputFile parquetInputFile = new ParquetInputFile(file);
+        return read(parquetInputFile);
+    }
 
-        String votable = metadata.getFileMetaData().getKeyValueMetaData().get("IVOA.VOTable-Parquet.content");
-        log.debug("VOTable: " + votable);
+    public VOTableDocument read(RandomAccessSource randomAccessSource) throws IOException {
+        RandomAccessSeekableInputStream parquetInputFile = new RandomAccessSeekableInputStream(randomAccessSource);
+        return read(parquetInputFile);
+    }
 
-        VOTableReader voTableReader = new VOTableReader();
-        VOTableDocument voTableDocument = voTableReader.read(votable);
+    public VOTableDocument read(InputStream inputStream) throws IOException {
 
-        try (org.apache.parquet.hadoop.ParquetReader<GenericRecord> reader = AvroParquetReader.<GenericRecord>builder(inputFile).build()) {
-            GenericRecord record;
+        File tempFile = null;
 
-            while ((record = reader.read()) != null) {
-                log.debug("Reading Record: " + record);
-                recordCount++;
+        try {
+            String fileName = "parquet-" + UUID.randomUUID();
+            tempFile = File.createTempFile(fileName, ".parquet");
 
-                // TODO: This is a temporary implementation to read the data
-                Map<String, Object> rowData = new HashMap<>();
-                for (Type field : schema.getFields()) {
-                    String fieldName = field.getName();
-                    rowData.put(fieldName, record.get(fieldName));
+            try (InputStream byteCountInputStream = new ByteCountInputStream(inputStream, BYTE_LIMIT);
+                OutputStream out = new FileOutputStream(tempFile)) {
```

Comment:
There are lots of ways to do this inputstream -> outputstream copy loop. There are some convenience methods in java.nio and for higher performance there is a class in ca.nrc.cadc.io named `MultiBufferIO` than implements double and triple buffer copy using one additional thread. I would recommend MultiBufferIO here.